File: src/main/scala/com/builtbroken/lib/render/wrapper/BlockRenderHandler.scala
package com.builtbroken.lib.render.wrapper

import com.builtbroken.lib.prefab.tile.spatial.{Tile, BlockTile}
import cpw.mods.fml.client.registry.{ISimpleBlockRenderingHandler, RenderingRegistry}
import net.minecraft.block.Block
import net.minecraft.client.renderer.RenderBlocks
import net.minecraft.item.ItemStack
import net.minecraft.world.IBlockAccess
import org.lwjgl.opengl.GL11._
import org.lwjgl.opengl.GL12._
import com.builtbroken.lib.transform.vector.Vector3

object BlockRenderHandler extends ISimpleBlockRenderingHandler
{
  val ID = RenderingRegistry.getNextAvailableRenderId()

  def renderInventoryBlock(block: Block, metadata: Int, modelID: Int, renderer: RenderBlocks)
  {
    if (block.isInstanceOf[BlockTile])
    {
      val tile = (block.asInstanceOf[BlockTile]).staticTile

      glEnable(GL_RESCALE_NORMAL)
      glPushAttrib(GL_TEXTURE_BIT)
      glPushMatrix()
      tile.renderInventory(new ItemStack(block, 1, metadata))
      glPopMatrix()
      glPopAttrib()
    }
  }

  def renderWorldBlock(access: IBlockAccess, x: Int, y: Int, z: Int, block: Block, modelId: Int, renderBlocks: RenderBlocks): Boolean =
  {
    var renderer: Tile = null

    /**
     * Try TileEntity rendering
     */
    val tile = access.getTileEntity(x, y, z)

    if (tile.isInstanceOf[BlockTile])
    {
      val spatial = tile.asInstanceOf[BlockTile]
      renderer = spatial.staticTile
    }

    /**
     * Try Block rendering
     */
    if (renderer == null && block.isInstanceOf[BlockTile])
    {
      val dummy = block.asInstanceOf[BlockTile]
      dummy.inject(access, x, y, z)
      renderer = dummy.getTile(access, x, y, z)
    }

    if (renderer != null)
    {
      return renderer.renderStatic(renderBlocks, new Vector3(x, y, z), 0);
    }

    return false
  }

  def shouldRender3DInInventory(modelId: Int) = true

  def getRenderId() = BlockRenderHandler.ID

}

File: src/main/scala/com/builtbroken/mc/client/json/render/block/BlockRenderHandler.java
package com.builtbroken.mc.client.json.render.block;

import com.builtbroken.mc.client.json.ClientDataHandler;
import cpw.mods.fml.client.registry.ISimpleBlockRenderingHandler;
import cpw.mods.fml.client.registry.RenderingRegistry;
import cpw.mods.fml.common.eventhandler.SubscribeEvent;
import net.minecraft.block.Block;
import net.minecraft.client.renderer.RenderBlocks;
import net.minecraft.world.IBlockAccess;
import net.minecraftforge.client.event.RenderWorldEvent;
import net.minecraftforge.common.MinecraftForge;
import org.lwjgl.opengl.GL11;
import org.lwjgl.opengl.GL12;

/**
 * @see <a href="https://github.com/BuiltBrokenModding/VoltzEngine/blob/development/license.md">License</a> for what you can and can't do with the code.
 * Created by Dark(DarkGuardsman, Robert) on 4/4/2017.
 */
public class BlockRenderHandler implements ISimpleBlockRenderingHandler
{
    public final static int ID = RenderingRegistry.getNextAvailableRenderId();

    //Render pass of the world, set by world render events
    private int pass = 0;

    public BlockRenderHandler()
    {
        ClientDataHandler.INSTANCE.addBlockRenderer("default", this);
        MinecraftForge.EVENT_BUS.register(this);
    }

    @Override
    public void renderInventoryBlock(Block block, int metadata, int modelID, RenderBlocks renderer)
    {
        GL11.glEnable(GL12.GL_RESCALE_NORMAL);
        GL11.glPushAttrib(GL11.GL_TEXTURE_BIT);
        GL11.glPushMatrix();
        //TODO render
        GL11.glPopMatrix();
        GL11.glPopAttrib();
    }

    @Override
    public boolean renderWorldBlock(IBlockAccess access, int x, int y, int z, Block block, int modelId, RenderBlocks renderBlocks)
    {

        //TODO get tile at location
        //TODO get tile's content ID
        //TODO load json render data for ID

        //TODO get tile's state ID
        //TODO generate list of data combinations (cache if possible)
        //      state.renderPass, state
        //TODO loop list trying to locate state
        //TODO render state
        //TODO IF rendered cache state for faster render time
        return false;
    }

    @Override
    public boolean shouldRender3DInInventory(int modelId)
    {
        return true;
    }

    @Override
    public int getRenderId()
    {
        return BlockRenderHandler.ID;
    }

    @SubscribeEvent
    public void postWorldRender(RenderWorldEvent.Post event)
    {
        pass = 0;
    }

    @SubscribeEvent
    public void preWorldRender(RenderWorldEvent.Pre event)
    {
        pass = event.pass;
    }
}
